File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        int count = 0; // Variable to store the number of subarrays that sum to k.
        int sum = 0; // Variable to store the cumulative sum of elements.

        unordered_map<int,int> mp;
        // Base case: there's one way to have a sum of 0 (no elements).
        mp[sum] = 1;
        for(int i = 0; i < n; i++)
        {
            sum += nums[i]; // Update the cumulative sum.
            int diff = sum-k;
            // If cumulativeSum - k i.e diff exists in mp, then a subarray ending at current index has a sum of k. We add the count of those occurrences to answer.
            if(mp.find(diff) != mp.end())
            {
                count += mp[diff];
            }
            // We then increment the count of cumulativeSum in our frequency map.
            mp[sum]++;
        }

        return count;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n = nums.length;
        Map<Integer, Integer> mp = new HashMap<>();
        int sum = 0;
        int count = 0;
        mp.put(sum,1);

        for(int i = 0; i < n; i++){
            sum += nums[i];

            int diff = sum - k;

            if(mp.containsKey(diff)){
                count += mp.get(diff);
            }

            mp.put(sum, mp.getOrDefault(sum, 0)+1);
        }

        return count;
    }
}